### PR TITLE
Multi-thread GroupReadsByUmi

### PR DESCRIPTION
**Pull request by nh13**

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala [hunk 1]
```diff
@@ -495,24 +498,41 @@ class GroupReadsByUmi
...
-        templatesByMi(tag).sortBy(t => t.name).flatMap(t => t.primaryReads).foreach(rec => {
-          out += rec
-        })
-      })
+    val tagFamilySizeCounters = new IterableThreadLocal(() => new NumericCounter[Int]())
+    val pool                  = new ForkJoinPool(threads, ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, true)
+
+    Iterator.continually(if (iterator.hasNext) takeNextGroup(iterator) else Seq.empty)
+      .takeWhile(_.nonEmpty)
+      .grouped(100000)
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala:507]: this is arbitrary

File: src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala [hunk 2]
```diff
@@ -495,24 +498,41 @@ class GroupReadsByUmi
...
+            // Count up the family sizes
+            val tagFamilySizeCounter = tagFamilySizeCounters.get()
+            templatesByMi.values.foreach(ps => tagFamilySizeCounter.count(ps.size))
+
+            templatesByMi.keys.toSeq.sortBy(id => (id.length, id)).flatMap { tag =>
+              templatesByMi(tag).sortBy(t => t.name).flatMap(t => t.primaryReads)
+            }
+          }
+          .seq
+          .iterator
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala:527]: never sure if this is needed, but here we are
```suggestion
          .seq
          .iterator
```

**Comment by tfenne** [comment]
I have a couple of non-trivial PRs for GroupReadsByUmi outstanding that I'd like to get merged before looking at this one.  Did you measure the performance difference - I'm curious how much this helps?

**Comment by nh13** [comment]
@tfenne yup, it’s easy to to rebase this later, and I’ll post performance metrics in the meantime.

File: src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala (hunk 1 above)
Comment (nh13) [reply src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala:507]: this needs to be tuned

File: src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala [hunk 3]
```diff
@@ -554,24 +557,41 @@ class GroupReadsByUmi
...
 
-      templatesByMi.keys.toSeq.sortBy(id => (id.length, id)).foreach(tag => {
-        templatesByMi(tag).sortBy(t => t.name).flatMap(t => t.primaryReads).foreach(rec => {
-          out += rec
-        })
-      })
+    val tagFamilySizeCounters = new IterableThreadLocal(() => new NumericCounter[Int]())
+    val pool                  = new ForkJoinPool(threads, ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, true)
+
+    Iterator.continually(if (iterator.hasNext) takeNextGroup(iterator) else Seq.empty)
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala:564]: will want to use https://github.com/fulcrumgenomics/commons/pull/76 when ready!

**Pull request closed without merging**